core/src/data_sources/data_source.rs

```diff
@@ -620,8 +698,38 @@ impl DataSource {
             };
 
             for (s, v) in chunk.into_iter().zip(v.into_iter()) {
-                e.push((offset, s, v));
-                offset += 1;
+                let mut hasher = blake3::Hasher::new();
+                hasher.update(s.as_bytes());
+                let text_only_hash = format!("{}", hasher.finalize().to_hex());
+
+                texthash2embedding.insert(text_only_hash, v);
+            }
+        }
+
+        // Ordered results with offset, stirng and vector
+        let embedded_vectors = splits
+            .iter()
+            .enumerate()
+            .map(|(i, s)| {
+                let mut hasher = blake3::Hasher::new();
+                hasher.update(s.as_bytes());
+                let text_only_hash = format!("{}", hasher.finalize().to_hex());
+                let r = match texthash2embedding.contains_key(&text_only_hash) {
+                    true => Ok((
+                        i,
+                        s.to_string(),
+                        texthash2embedding.get(&text_only_hash).unwrap(),
+                    )),
+                    false => Err(anyhow!("Chunk not found in cache")),
+                };
+                r
+            })
+            .collect::<Vec<_>>();
+
+        for r in embedded_vectors.iter() {
```

Comment:
I am sure there is a more Rusty way of handling an error here but I could not figure it out.
I tried try_collect::<>() but the stdlib version of it warns me that it is an experimental feature.

Reply:
Yes we definitely can do better I think.

Reply:
Happy to help once the other comments are addressed :+1:

Reply:
ok, I tried again and asked chatGPT, could not find a better solution. LMK what you have in mind.

Reply:
I think we can do it at the point of execution, will give it a try if fine with you?